rewrite-java/src/main/java/org/openrewrite/java/cleanup/DeencapsulateObjects.java

```diff
@@ -0,0 +1,149 @@
+import org.openrewrite.java.JavaVisitor;
+import org.openrewrite.java.tree.J;
+import org.openrewrite.java.tree.J.Identifier;
+import org.openrewrite.java.tree.JavaType;
+
+public class DeencapsulateObjects extends Recipe {
+  @Override protected TreeVisitor<?, ExecutionContext> getVisitor() {
+    return new Visitor();
+  }
+
+  public static class Visitor extends JavaIsoVisitor<ExecutionContext> {
+
+    static String degetterify(String input) {
+      String getterRemoved = input.replaceAll("^get([A-Z])", "$1");
+      // Fastest way to lowercase the first character of a string
+      // https://stackoverflow.com/a/31735171
+      char[] c = getterRemoved.toCharArray();
+      c[0] = Character.toLowerCase(c[0]);
+      return new String(c);
+    }
+
+    @Override public J.MethodDeclaration visitMethodDeclaration(        J.MethodDeclaration method,        ExecutionContext notused    ) {
+      AtomicReference<J.MethodDeclaration> modifiedMethod =
+          new AtomicReference<>(super.visitMethodDeclaration(method, notused));
+      AtomicBoolean wasModified = new AtomicBoolean(false);
+
+      modifiedMethod.get().getParameters()
+          .stream()
+          .filter(i -> i instanceof J.VariableDeclarations)
+          .map(i -> (J.VariableDeclarations) i)
+          .forEach(it -> {
+
+            Identifier parameter = it.getVariables().get(0).getName();
+            FindParameterUsages usagesFinder = new FindParameterUsages(parameter);
+            usagesFinder.visit(modifiedMethod.get(), modifiedMethod.get());
+            Set<Identifier> uniqueUsages = usagesFinder.uniqueUsages;
+            if (uniqueUsages.size() == 1) {
+              wasModified.set(true);
+              // fix param
+              Identifier usage = uniqueUsages.stream().findFirst().get();
+
+              String replacedParameterName = degetterify(usage.getSimpleName());
+              modifiedMethod.set(new ReplaceParameterWithCaller(
+                  parameter,
+                  replacedParameterName,
+                  usagesFinder.returnType
+              ).visitMethodDeclaration(modifiedMethod.get(), null));
+
+              // find and replace usages
+              modifiedMethod.set((J.MethodDeclaration) new ReplaceInvocationsWithParameter(
+                  replacedParameterName,
+                  usage
+              ).visitMethodDeclaration(modifiedMethod.get(), null));
+            }
+          });
+      if (wasModified.get()) {
+        // find invocations of this method and change to the new signature
+        //      val classname = cursor.firstEnclosingRequired()
+        //      val params = method.params.params
+        //      val signature = classname + "." + method.name.printTrimmed() + "(" + params + ")"
+        //      println(signature)
+        // andThen()
+      }
+      return modifiedMethod.get();
+    }
+  }
+
+  public static class ReplaceInvocationsWithParameter extends JavaVisitor<Void> {
```

Comment:
I think conceptually, this can be done with two visitors. The first is looking for method declarations where we find one of the DTO patterns. (This could then store a marker on the method declaration that includes the parameter(s) that are being replaced along with the getter for each parameter)

Then the second visitor is the one that acts on the marker, replacing both the parameter and the invocation in the method body.